### ISSUE TITLE
[multiple] Serialization of rules

### ISSUE THREAD

**Issue by tabatkins** [w3c/csswg-drafts#4828]
I just merged a PR (<https://github.com/w3c/css-houdini-drafts/pull/980>) specifying how to serialize an `@property` rule. I wasn't sure whether it should all be serialized on one line (as the PR specified, and now in the draft), or if they should be indented on a newline, so I went looking for precedents, and it doesn't appear we have any.

Afaict, no spec defines how to serialize rules *at all*: not CSSOM, nor any spec that defines its own at-rules, like Conditional or Animations.

I presume we want to specify this so we can be consistent, and can test rule serialization? How should it work? Where should such a definition live - individual specs, or in CSSOM, or general rules in CSSOM with explicit hooks that individual specs have to define for serialization of preludes and which descriptors to include?

**Comment by tabatkins** [comment]
Agenda+ to bring the issue into general awareness; I don't expect us to decide these questions in the call.

**Comment by svgeesus** [comment]
Yeah it would be good to get some discussion on this, so for example fonts 4 can say how to serialize the at-rules it has (or I can give input on what goes in CSSOM. However, we recently decided to move color serialization from CSSOM to color 4, so...)

**Comment by tabatkins** [comment]
So it turns out browser's *aren't* consistent:

[testcase](http://livedomviewer.com/?%3C!DOCTYPE%20html%3E%0A%3Cstyle%3E%0A%40font-face%20{%0A%20font-family%3A%20foo%3B%20%20src%3A%20url(foo)%3B%0A%20font-weight%3A%20bold%3B%0A}%0A%3C%2Fstyle%3E%0A%3Cscript%3E%0Aw(document.styleSheets[0].cssRules[0].cssText)%3B%0A%3C%2Fscript%3E):
```html
<!DOCTYPE html>
<style>
@font-face {
 font-family: foo; src: url(foo);
 font-weight: bold;
}
</style>
<script>
w(document.styleSheets[0].cssRules[0].cssText);
</script>
```

Chrome puts all of it on one line. Firefox puts each descriptor on its own line.

@emilio You expressed during the call a preference for single-line, I think? Are you okay with specifying that and getting Firefox to fix?

**Comment by Loirooriol** [comment]
But with `@keyframes` both browsers use multiple lines (Chromium with indentation, Firefox without)
[testcase](http://livedomviewer.com/?%3C!DOCTYPE%20html%3E%0A%3Cstyle%3E%0A%40keyframes%20anim%7Bfrom%7Bcolor%3Ablue%7Dto%7Bcolor%3Ared%7D%7D%0A%3C%2Fstyle%3E%0A%3Cscript%3E%0Aw(document.styleSheets%5B0%5D.cssRules%5B0%5D.cssText)%3B%0A%3C%2Fscript%3E)

**Comment by emilio** [comment]
ew, what a mess. I think changing style rules is a no-go, so I'd be happy to do everything in one line if other browsers are happy with that too.

**Comment by tabatkins** [comment]
Ah, yeah, both Chrome and Firefox do style rules on a single line, and I agree that's the most common case and most likely to be depended on, so we should go with that.

**Comment by emilio** [comment]
https://drafts.csswg.org/cssom/#serialize-a-css-rule does have some text that matches Firefox and Chrome. Spec for some rules is missing and we should add it though.

**Comment by cdoublev** [comment]
> **CSSImportRule**
>
>The result of concatenating the following:
> 1. The string `@import` followed by a single SPACE (U+0020).
> 2. The result of performing serialize a URL on the rule’s location.
> 3. If the rule’s associated media list is not empty, a single SPACE (U+0020) followed by the result of performing serialize a media query list on the media list.
> 4. The string `;`, i.e., SEMICOLON (U+003B).

Serializing `@import` currently requires serializing its location as an `<url>` (even when specified as a `<string>`) and serializing its *associated media list* with *serialize a media query list*, but `CSSImportRule` does not have a property that is a list of component values matching `<media-query-list>`.

I tested in Chrome/Firefox, and both seem to use `CSSImportRule.media`, which is a live reference to `CSSImportRule.styleSheet.media`, meaning it may not correspond to the value authored in the style sheet for `@import` if it has been updated with `CSSImportRule.styleSheet.media.mediaText` or `CSSImportRule.styleSheet.media.appendMedium()`. Also note that currently there is nothing preventing the media list from being updated, while you can't update the imported style sheet rules.

Suppose that you specify `supports(display: unsupported)` as an import condition (supports condition in `@import` do not seem to be already supported in Chrome/Firefox, which parse/serialize it to `not all`), then UA *must return null for the import rule’s `CSSImportRule.styleSheet` value*. Obviously, I should I keep/use an internal reference to the prelude value authored in the style sheet for `@import`.

Please let me know if you think there is anything wrong/unexpected regarding my observations.
